src/boot/measure.c

```diff
@@ -464,6 +525,235 @@ static int verb_calculate(int argc, char *argv[], void *userdata) {
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                NULL,
+                                TPM2_SE_TRIAL,
+                                &symmetric,
+                                TPM2_ALG_SHA256,
+                                &session_handle);
+                if (rc != TSS2_RC_SUCCESS) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                            "Failed to open session in TPM: %s", sym_Tss2_RC_Decode(rc));
+                        goto finish;
+                }
+
+                /* Generate a single hash value from the PCRs included in our policy. Given that that's
+                 * exactly one, the calculation is trivial. */
+                TPM2B_DIGEST intermediate_digest = {
+                        .size = SHA256_DIGEST_SIZE,
+                };
+                assert(sizeof(intermediate_digest.buffer) >= SHA256_DIGEST_SIZE);
+                sha256_direct(p->value, p->value_size, intermediate_digest.buffer);
+
+                TPM2_ALG_ID tpmalg = convert_evp_md_name_to_tpm2_alg(p->md);
+                if (tpmalg == TPM2_ALG_ERROR) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Unsupported PCR bank");
+                        goto finish;
+                }
+
+                TPML_PCR_SELECTION pcr_selection;
+                tpm2_pcr_mask_to_selection(1 << TPM_PCR_INDEX_KERNEL_IMAGE, tpmalg, &pcr_selection);
+
+                rc = sym_Esys_PolicyPCR(
+                                c.esys_context,
+                                session_handle,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                &intermediate_digest,
+                                &pcr_selection);
+                if (rc != TSS2_RC_SUCCESS) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                            "Failed to push PCR policy into TPM: %s", sym_Tss2_RC_Decode(rc));
+                        goto finish;
+                }
+
+                _cleanup_(Esys_Freep) TPM2B_DIGEST *pcr_policy_digest = NULL;
+                rc = sym_Esys_PolicyGetDigest(
+                                c.esys_context,
+                                session_handle,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                &pcr_policy_digest);
+                if (rc != TSS2_RC_SUCCESS) {
+                        r = log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                            "Failed to get policy digest from TPM: %s", sym_Tss2_RC_Decode(rc));
+                        goto finish;
+                }
```

Comment:
The newer libpolicy on tpm2-tss will calculate all this without round tripping to the TPM. Probably too new for usage here, but calculating this by hand is pretty easy:
`policyDigestnew ≔ HpolicyAlg(policyDigestold || TPM_CC_PolicyPCR || pcrs || digestTPM)`

Where:
- policyDigestold is 32 0's
- TPM_CC_PolicyPCR is constant `TPM2_CC_PolicyPCR` from `tss2_tpm2_types.h` aka value 0x017f
- pcrs: libmu marshalled `TPML_PCR_SELECTION`.
- digestTPM: The expected digest of all the PCRs

Sample code can be found in this function:
  - https://github.com/tpm2-software/tpm2-tss/blob/cdeba44f2273873ed73f71c75dfed9738ab36691/src/tss2-fapi/ifapi_policy_calculate.c#L84

You'd be able to make setup TPM agnostic. Although you could just fire up a simulator for all of this.

Reply:
Ah interesting. We can certainly look into calculating this without relying on a TPM. Doing this without any TPM would be excellent of course. For the testsuite we are currently using swtpm for this, and my thinking so far was that people can use that for systemd-measure too on their build systems.

But you are right, if it's easy to hash the values without the TPM we should really do this.

Reply:
is this going to be addressed in this PR?

Reply:
I'd do that later.